pkg/backup/restore_job.go

```diff
@@ -154,48 +157,63 @@ func rewriteBackupSpanKey(
 	return newKey, nil
 }
 
-func restoreWithRetry(
-	restoreCtx context.Context,
-	execCtx sql.JobExecContext,
+type retryState int
+
+const (
+	initialRetryState retryState = iota
+	secondaryRetryState
+)
+
+type restoreRetrier struct {
+	execCtx sql.JobExecContext
+	resumer *restoreResumer
+
+	initialRetryOpts      retry.Options
+	secondaryRetryOpts    retry.Options
+	policySwitchThreshold float32
+
+	// Retry state fields
+	retry              retry.Retry
+	currRetryState     retryState
+	prevPersistedSpans jobspb.RestoreFrontierEntries
```

Comment:
Do we really need to keep two copies of this potentially large list of keys bytes around?

Could we just have the progress update code be responsible updating a timestamp or fraction we can check against?

Reply:
Oh right not both copies are needed, just the one.

Reply:
It does seem like this checkpoint is only used in this part of the code, so actually storing the spans may not be necessary. Although I think if we do want to update this, it would be outside of this PR.